From e90bcefab50c6297d9ab2ab99cdc5b0622e7beb2 Mon Sep 17 00:00:00 2001
From: Christophe Chapuis <chris.chapuis@gmail.com>
Date: Sat, 25 Nov 2023 11:18:06 +0100
Subject: [PATCH] WebOSSurfaceItem: close Wayland client fallback on Close
 request

When external native apps, like SDL apps, are started on LuneOS, they
don't implement the webOS Wayland extension. This means no shell
surface is created, and we have no means of closing the window.
In this case, just request the Wayland client to close, as a fallback.

Signed-off-by: Christophe Chapuis <chris.chapuis@gmail.com>
Upstream-Status: Pending
---
 modules/weboscompositor/webossurfaceitem.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/modules/weboscompositor/webossurfaceitem.cpp b/modules/weboscompositor/webossurfaceitem.cpp
index dc696aa..caaf5fd 100644
--- a/modules/weboscompositor/webossurfaceitem.cpp
+++ b/modules/weboscompositor/webossurfaceitem.cpp
@@ -1067,7 +1067,10 @@ void WebOSSurfaceItem::close()
         sendCloseToGroupItems();
         m_shellSurface->close();
     } else {
-        qWarning() << "No webos shell surface exist, cannot close";
+        qWarning() << "No webos shell surface exist, will close entire client !";
+        if (surface() && surface()->client()) {
+            surface()->client()->close();
+        }
     }
 }
 
-- 
2.43.0

